### ISSUE TITLE
Safari Support

### ISSUE THREAD

**Issue by aileen-shih** [jimmywarting/StreamSaver.js#69]
## Update:
(2019-06-18)  by @jimmywarting

Want to keep this issue open until safari gets "real" support for downloading something that is generated by service worker. (even if it's not a stream) think this as some good reference to some safari bugs and background to what's wrong with safari

cuz we eventually need to remove the special edge case - hard coded scenario that isn't such a graceful degradation when it comes to Safari

https://github.com/jimmywarting/StreamSaver.js/blob/c92eaf7373a4e46236733efe7fe4509f874a6560/StreamSaver.js#L17

I don't know yet how we should detect support for Safari as they yield false/positive atm, maybe unwillingly do UA sniffing instead of feature detection... but that will be an task for later when safari can really save something from service worker

## Original issue

First of all, just want to thank you for this great work, it works beautifully in Chrome!

I see Safari's missing support for serviceWorker is already supported in their latest version 11.1. https://caniuse.com/#search=serviceWorker
So I tested in v11.1.2 Safari but it is not working, am I missing something?
(for testing, I removed support check for WritableStream)

**Comment by jimmywarting** [comment]
safari supports both ReadableStream and service worker. But it seems like they have some bug when it comes to assemble a response with a stream and returning it.

```js
stream = new ReadableStream({
  start(controller) {
    controller.enqueue(uint8array)
  }
})
response = new Response(stream, { headers })
respondWith(response)
```

**Comment by alippai** [comment]
What about Safari 12 now? Did anybody test it? :P

**Comment by alippai** [comment]
May this statement "The implementation does not support ReadableStreamBYOBReader." cause the problem? https://webkit.org/status/#feature-readable-streams

**Comment by jimmywarting** [comment]
I tested safari like 3 days ago. My plan later is not to use PostMessage but to use the fetch api once `request.body` is available in as a readable stream. Safari was first to implement it in `request`

The idea is just to pipe the request back to response
something like this:

```js
onfetch = evt => {
  const { request } = evt
  const response = new Response(request.body, {
    headers: new Headers(request.headers)
  })
  evt.respondWith(response)
}
```

It worked just fine to stream the request back to the response but whenever i added the `content-disposition` header if would just fail with a 404 result. So i tried other hack: When `Content-Type` is set to something that the browser can't handle like a `application/zip` it will download it also using the url's filename if no attachment header is available. but it was same problem there as well :( 

So i tried with simple `Content-Type: text/html` it works, it's displayed properly... now add download attribute to the link and press it - fail 😢 

The problem don't seem to be BYOB reader or any readable stream. It seems to me whenever safari would like to save something. it switches techniques and tries fetch the url directly instead of going throught with what the service worker respond with.

A simple example such as this dose not work:

```js
event.respondWith(new Response('hi', {
  headers: {
    Content-Disposition: attachment; filename="world.txt"
  }
}))
```

Then we are not even using streams... the result is you download some file but it's not the content you responded with. it's the url behind service worker. 

somebody should report this...

**Comment by alippai** [comment]
Added as a quick comment here: https://bugs.webkit.org/show_bug.cgi?id=167341
I'm not sure, the issues are connected, however Frédéric Wang's comment could cause this behavior.

**Comment by jimmywarting** [comment]
Thanks the "decide policy for navigation action" seems to be in wrong order...?

**Comment by johnozbay** [comment]
Just wanted to bring some good news that the [bug report](https://bugs.webkit.org/show_bug.cgi?id=167341) @alippai is referencing is now resolved, and Safari should now support download tags, along with many other fantastic features in 12.2b, some of which are listed [here](https://twitter.com/mhartington/status/1089292031548145666). (looks like service workers are removed from the experiments list)

Discovered StreamSaver just this week, and still in the process of understanding how it works. So I'm not entirely sure if these new bits of Safari features would help cure the symptoms mentioned in this thread, but figured I'll post here, and more info won't hurt.

Thanks a lot for authoring this fantastic piece of code @jimmywarting! ✌🏻

**Comment by jimmywarting** [comment]
That's grate news. Still doe - i don't have high hopes that it's fixed yet
that bug report and twitter post was for iOS.

Desktop Safari have had the service worker & download attribute for quite some time now. Even the Writable & Readable Stream wish is the most important parts of streamsaver to make everything work

But lets have hope that the changes for mobile safari affected the desktop browser to handle downloading files better

----

> still in the process of understanding how it works

In a nutshell we create a Service Worker and mimic what a server would do when you want to save something. 1) Add some response header to tell browser to save it instead. 2) Send a few chunks and then some more by using streams 
```js
evt.respondWith(new Response(new ReadableStream(...), { headers }))
```
How we get the chunks to the service worker is black magic (we use postMessage to transfer data)
(But i guess you had already figure out the basic of how streamsaver works by now)

**Comment by johnozbay** [comment]
That makes perfect sense! And I now understand it a lot better! Thank you so much for the mini-intro. 
I'll keep digging further, and keep this thread updated if I can find more info from Apple. Cheers! ✌🏻

**Comment by biscoe916** [comment]
@johnozbay Running into this now. Did you find anything new out?

**Comment by johnozbay** [comment]
@biscoe916 Nothing new as far as I know. I'm actively monitoring progress on this. Will try my best and keep this thread up to date as much as I can. – If I had to take a guess I'd say it won't be for another few months or so. iOS Safari folks are adding in tons of PWA / service worker etc. features to Safari betas every few weeks nowadays, presumably in prep for the next big iOS release. 

If you wish to keep a personal tap open on this follow along : [Jonathan Davis](https://twitter.com/jonathandavis) chances are that twitter account will light up like a christmas tree when this (and many other features) arrive.

**Comment by jimmywarting** [comment]
the problem is that safari detects `Content-Disposition` attachment header, starts to ignore everything that comes from the service worker and download everything as if it came from the server. The result is that you get the response you normally would get without a service worker in place. 

It already has all the pieces needed (fetch, service worker, readable streams) but safari dose something strange when it comes to download stuff - almost like it switches technology, like using some builtin download manager that are unaware of the service worker or something

**Comment by jimmywarting** [comment]
Every other evergreen browser now support saving something from the service worker except for safari...

To tackle Safari (and maybe other older browser) I have decided to do a graceful degradation to build a blob from the stream and simply use a[download] to save it.

**Comment by jimmywarting** [comment]
Maybe it's time to test safari again if it can save something from a service worker
finger crossed  🤞 

Don't have big hopes for it to be fixed there has only been 6 months since i looked at it the last time.

**Comment by jimmywarting** [comment]
still don't work 😞

**Comment by jimmywarting** [comment]
This looks more like the correct issue related to our bug
https://bugs.webkit.org/show_bug.cgi?id=182848

It's says fixed, but is it released?

**Comment by johnozbay** [comment]
@jimmywarting Looks like it's in the nightly. (Haven't tested it yet) You can give it a try with the Safari Technology Preview tho : http://developer.apple.com/safari/technology-preview/ 

Most of these changes will likely go live with the iOS 13 release.

**Comment by jimmywarting** [comment]
I tried Safari Technology Preview  as well...

**Comment by ninjasource** [comment]
Thanks for checking again @jimmywarting. For anyone else reading this, if you absolutely have to get this technology working on iOS then Jimmy's library works as expected on the iOS version of Firefox available on the app store.

**Comment by jimmywarting** [comment]
oh, btw new tracking bug to follow: https://bugs.webkit.org/show_bug.cgi?id=201580

**Comment by balintx99** [comment]
What I see is, that now Safari asks the user if he/she wants to download the file, but in the end only a 9 kb dummy file gets through. I read above that there is a fallback far Safari, is this automatic? It doesnt seems to work in my case.

Huge thanks for this package, it is an absolute lifesaver🎉

**Comment by TheMrZZ** [comment]
Still no movements on this issue. A real shame... Safari is really the new IE.

**Comment by eligrey** [comment]
Safari (Technology Preview) 14.1 supports WritableStream afaik, so this issue can now be revisited.

**Comment by jimmywarting** [comment]
Sweet!

**Comment by christianwengert** [comment]
what would be the entry point to test a recent Safari Version to see if StreamSaver also works with the WritableStream of Safari?
I would happily try out but do notquite know where to start

**Comment by jimmywarting** [comment]
@christianwengert  I have had a long time to try and figure that out. But haven't thought about something yet.

The issue is not whatever or not safari supports WritableStreams (as @eligrey mention) ...it can be polyfilled and isn't strictly necessary for StreamSaver to do what it needs to do.

---

Safari already has support for service worker, readable+writable streams (and everything else to make StreamSaver work). That's why it's so hard to test something that isn't "really" broken (...in javascript). We might just have to do browser detection instead 😞 

---

The issue relies on wether or not the safari has addressed this bad/partially fix: https://bugs.webkit.org/show_bug.cgi?id=182848
They solved service worker downloads by aborting the request entirely and avoid the hole `fetchEvent.respondWith(...)` by repeating and making a "real" request to the destinated target and avoid the service worker intercepter all together.

> "This patch is not fully fixing the issue, it only allows the network process to redo the download by triggering another network request.
> Any service worker generated content will not be downloaded."

The new issue to fix it is: https://bugs.webkit.org/show_bug.cgi?id=202142 (and it haven't been resolved yet)


---

A easy way to (manually) test if it works is to register a service worker and intercept a network request with a custom response

```js
globalThis.onfetch(evt => {
  if (evt.request.url !== some_url) return

  evt.respondWith(new Response('hi', {
    headers: {
      'Content-Disposition': 'attachment; filename="world.txt"',
      'Content-Type': 'application/octet-stream'
    }
  }))
})
```

if the downloaded file says `hi` in a file called `world.txt` then it works! As of right now it will just make a request to `some_url` that probably don't exist and download a 404 file content instead. 

I haven't bothered to test it out since the webkit bug is still in the "new" state and not in "fixed"

**Comment by christianwengert** [comment]
Ok I installed a serviceworker as yours above 
```
if(url.endsWith('/bla')) {

    event.respondWith(new Response('hi', {
      headers: {
        'Content-Disposition': 'attachment; filename="world.txt"',
        'Content-Type': 'application/octet-stream'
      }
    }))

  }
```


and have a function in the main thread as follows:

```
document.addEventListener('dblclick', function(event) {
    fetch('/bla').then(r=>console.log(r))
})
```



But is does not download anything :(

I just get this as output:
```
[Log] Response {type: "basic", url: "http://localhost:8008/bla", redirected: false, status: 200, ok: true, …}
```


So no download at all (again!). Am I missing something or is this due to the above mentionned bug?

Same in Safari 14.1 and Safari 14.2  Tech Preview

**Comment by jimmywarting** [comment]
you can't download something if u are using ajax to fetch something (a browser navigation needs to happen) or a iframe needs to be opened.

Besides you are just logging the response, try logging the body 
```js
fetch('/bla').then(r=>r.text()).then(console.log)
```

Also you could try a simple `<a href="/bla">download</a>`

--- 

tip: you can always write the language to do syntax color, example:

\```js
code
\```

**Comment by christianwengert** [comment]
Ok I tried the above with a ```<a href="/bla">download</a>``` as well as with a ```<a href="/bla" download="bla.txt">download</a>```

The call in the service workers gets intercepted and the call

```js
    event.respondWith(new Response('hi', {
      headers: {
        'Content-Disposition': 'attachment; filename="world.txt"',
        'Content-Type': 'application/octet-stream'
      }
    }))
```

returns. 

This saves a file (!) but the content is not "hi" as expected, but just the server response:
```
Not Found

The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.
```

The same happens BTW if I try the same with an iframe

```js
const iframe = document.body.appendChild(document.createElement('iframe')) as HTMLIFrameElement
iframe.src = '/bla'
```

It looks like the bug is still valid :(

**Comment by christianwengert** [comment]
Also if I log the body of the javascript ```fetch``` call, I also get as contents the ```Not found``` content

**Comment by jimmywarting** [comment]
...Sadly as I expected

**Comment by christianwengert** [comment]
@jimmywarting Thanks for giving me some help to test this. I will gladly help again **if** there are changes on this subject...lets hope for safari15

**Comment by christianwengert** [comment]
Little update for everyone interested: Safari 15 still does not work.

**Comment by christianwengert** [comment]
Latest Tech Preview (137) from Safari: 

If I have in the sw.js:

```

if(url.indexOf('/bla') >= 0) {
    console.log('blaing')
      return event.respondWith(new Response('hello there', {
        headers: {
          'Content-Disposition': 'attachment; filename="world.txt"',
          'Content-Type': 'application/octet-stream',
          "Cache-Control":  "must-revalidate",
          "Pragma": "must-revalidate"

        }
      }))

  }
```

and a download button ```<a href="/bla" id="webkit-bug">download</a>```

and a JS-File 
```
document.getElementById('webkit-bug').addEventListener('click', (e)=>{
        e.preventDefault()
        fetch('/bla').then(r=>r.text()).then(console.log)
    })
```

I get the following:
**hello there** in the console +1 


As a sidenote: If I do not do the event listener and just click on the download button (no JS) I still get a download of a html file with page not found content.

However I cannot test the Tech Preview on my App, because there are currently other safari issues

**Comment by jimmywarting** [comment]
This is how it has been for a while now, safari likes to cancel the request and re-download the resource in a background thread and skipping the service worker entirely

**Comment by christianwengert** [comment]
aha, so the whole bugfixing from the webkit team did not advance this topic at all ?

**Comment by jimmywarting** [comment]
hmm, don't know yet... haven't tested current TP

**Comment by jimmywarting** [comment]
> > There’s also new support for allowing users to download files generated by a ServiceWorker.
> https://webkit.org/blog/12445/new-webkit-features-in-safari-15-4/
> 
> Safari 15.4 release notes mention the handling of files in ServiceWorker.
> 
> Until now, I think that Safari could not process files that were larger than the amount of memory and were divided.
> 
> Has safari it improved in Safari 15.4??

_Originally posted by @ogawa0071 in https://github.com/jimmywarting/StreamSaver.js/issues/272#issue-1171183391_

**Comment by reinos** [comment]
Any update on this issue?

**Comment by reinos** [comment]
I've posted that the issue is possible fixed in Safari, but im still not sure. Currently we dropped Stream-saver for our project. We now setup a service-worker and tried to push a stream to the worker... FAIL... We tried to setup a MessageChannel between the service-worker and the app, and push the stream via the messageChannel to the service-worker... FAIL... 

Bottom line; it seems that Safari did some fixes on the part of Service-worker and transferable's but I have still the feeling that it is not working 100%. 

For now we have a service-worker running and construct the Stream inside the service-worker. So, we are not pushing a stream to the service-worker but just the values to construct the stream inside the service-worker.

Still like the concept of Stream-saver as it is trying to fix the gap between the browsers. However, the current state of transferable streams in the browser landscape is far from perfect! I hope and wish that this should be fixed soon as the concept (and working version in Chrome) is awesome!

**Comment by jimmywarting** [comment]
The "Gap" you are talking about is kind of being dealt with using [File system access](https://github.com/WICG/file-system-access/). it is far from perfect cuz it's quite restrict and only being shipped in blink, firefox & safari are a bit against it...

**Comment by pypt** [comment]
The newest version of Safari (Version 18.1 (20619.2.8.11.10)) appears to work fine.

I've tried setting `useBlobFallack` to `false` here:

https://github.com/jimmywarting/StreamSaver.js/blob/5b372f7ba5f1e82ef80e2466023930ee2c616a28/StreamSaver.js#L23

and it appears to be downloading my dynamically created file:

<img width="395" alt="Screenshot 2024-10-30 at 23 56 02" src="https://github.com/user-attachments/assets/a1ced9d3-9bd4-4b4e-bb21-b5094d268fd9">

`examples/plain-text.html` works too.

**Comment by pypt** [comment]
I've disabled the Safari hack:

https://github.com/pypt/StreamSaver.js-safari/commit/5c475048a77b2b0eab3a6c57abada2b9661c65b9

and tried out [`examples/plain-text.html`](https://pypt.github.io/StreamSaver.js-safari/examples/plain-text.html) with various Safari versions.

Safari version          | macOS version             | Works?
----------------------- | ------------------------- | ------
18.1 (20619.2.8.11.10)  | Sequoia 15.1 (24B83)      | ✅ Yes
17.6 (19618.3.11.11.5)  | Sonoma 14.7.1 (23H222)    | ✅ Yes
16.6 (18615.3.12.11.2)  | Ventura 13.7.1 (22H221)   | ✅ Yes
15.6.1 (17613.3.9.1.16)	| Monterey 12.7.6 (21H1320) | ❌ No

So, it appears that the Safari hack is only needed for Safari versions before 16.6.

**Comment by ogawa0071** [comment]
@pypt Safari gives me an error if the file size is too large. Will it work with files over 10GB?

**Comment by pypt** [comment]
@ogawa0071 which version of Safari?

**Comment by pypt** [comment]
@ogawa0071 I've tried my [patched version of `examples/plain-text.html`](https://pypt.github.io/StreamSaver.js-safari/examples/plain-text.html) with Safari 18.1 (20619.2.8.11.10), wrote ~12 GB worth of Loren Ipsum, and seemed to work fine:

<img width="817" alt="Screenshot 2024-10-31 at 15 03 45" src="https://github.com/user-attachments/assets/6a115897-f104-4a08-af19-89c72cf863b0">

Haven't tried the older browsers though, but I'd assume that they would work as nothing gets buffered in RAM.

**Comment by ogawa0071** [comment]
@pypt 
That's great! Just what I wanted!!!
What API did Safari support or what bug did it fix?

**Comment by tancgo** [comment]
@pypt   [https://pypt.github.io/StreamSaver.js-safari/examples/plain-text.html](url)， In Safari, clicking the abort button does not cancel the download task，are there any solutions?  thx

**Comment by pypt** [comment]
@tancgo which version of Safari are you using?
